lib/triton-tags.js

```diff
@@ -0,0 +1,185 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+/*
+ * Copyright 2016, Joyent, Inc.
+ */
+
+/*
+ * In Triton some per-VM configuration is controlled via special structured
+ * tags on the VM's "tags" object. They are all prefixed with "triton.".
+ * Let's call them "Triton tags".
+ *
+ * Triton tags are all optional.
+ *
+ * - `triton.cns.disable` (boolean): Can be set on a VM to tell the CNS service
+ *   to not serve records for this VM.
+ * - `triton.cns.services` (string): Comma-separated list of DNS-name strings
+ *   for the CNS service.
+ * - `triton.cns.reverse_ptr` (string): DNS reverse pointer for this VM. Used
+ *   by the CNS service.
+ */
+
+var assert = require('assert-plus');
+var format = require('util').format;
+
+
+/*
+ * For now, using the more limited labels allowed by RFC1123. RFC2181 supercedes
+ * 1123, but the broader range of characters can sometimes cause problems with
+ * other systems (e.g. see the underscore in RFC5321).
+ */
+var DNS_NAME_RE = /^[a-z0-9][a-z0-9\-]{0,62}(?:\.[a-z0-9][a-z0-9\-]{0,62})*$/i;
+
+const TRITON_TAG_PREFIX = 'triton.';
+
+var typeFromKey = {
+    'triton.cns.services': 'string',
+    'triton.cns.disable': 'boolean',
+    'triton.cns.reverse_ptr': 'string'
+};
+
+/*
+ * Validator functions take a `val` (already checked to be of the appropriate
+ * type per `typeFromKey`) and should throw an `Error` if invalid. The don't
+ * return a value.
+ */
+var validatorFromKey = {
+    'triton.cns.services': function validateTritonCnsServices(val) {
+        assert.string(val, 'val');
+
+        var unsafes = [];
+        var fqdns = val.split(',');
+        for (var i = 0; i < fqdns.length; i++) {
+            var fqdn = fqdns[i];
```

Comment:
This logic looks like it came from VMAPI and isn't quite correct. I have a change I'll be merging in (thanks to @arekinath) to VMAPI shortly and we'll want that same logic here if it must be copied. The new change allows for SRV records to be used with CNS (e.g. triton.cns.services=jabber:5223).

Should we think about a way to better re-use code?